crates/matrix-sdk/src/encryption/mod.rs

```diff
@@ -227,6 +230,121 @@ impl CrossProcessLockStoreGuardWithGeneration {
     }
 }
 
+/// A stateful struct remembering the cross-signing keys we need to upload.
+///
+/// Since the `/_matrix/client/v3/keys/device_signing/upload` might require
+/// additional authentication, this struct will contain information on the type
+/// of authentication the user needs to complete before the upload might be
+/// continued.
+///
+/// More info can be found in the [spec].
+///
+/// [spec]: https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3keysdevice_signingupload
+#[derive(Debug)]
+pub struct CrossSigningResetHandle {
+    client: Client,
+    upload_request: UploadSigningKeysRequest,
+    signatures_request: UploadSignaturesRequest,
+    auth_type: CrossSigningResetAuthType,
+}
+
+impl CrossSigningResetHandle {
+    /// Get the [`CrossSigningResetAuthType`] this cross-signing reset process
+    /// is using.
+    pub fn auth_type(&self) -> &CrossSigningResetAuthType {
+        &self.auth_type
+    }
+
+    /// Continue the cross-signing reset by either waiting for the
+    /// authentication to be done on the side of the OIDC issuer or by
+    /// providing additional [`AuthData`] the homeserver requires.
+    pub async fn auth(&self, auth: Option<AuthData>) -> Result<()> {
+        let mut upload_request = self.upload_request.clone();
+        upload_request.auth = auth;
+
+        // TODO: Do we want to put a limit on this infinite loop? 🤷
```

Comment:
Yes 🙂

Reply:
Which kind of limit, time-based or attempt based? The auth team argued that there's no need for a limit, which I can see since the user can just cancel the request/future if they would like to abort this.